sound/soc/sof/sof-audio.c

```diff
@@ -257,6 +257,79 @@ static int sof_setup_pipeline_connections(struct snd_sof_dev *sdev,
 	return 0;
 }
 
+static void
+sof_unprepare_widgets_in_path(struct snd_sof_dev *sdev, struct snd_soc_dapm_widget *widget)
+{
+	const struct sof_ipc_tplg_ops *ipc_tplg_ops = sdev->ipc->ops->tplg;
+	const struct sof_ipc_tplg_widget_ops *widget_ops = ipc_tplg_ops->widget;
+	struct snd_sof_widget *swidget = widget->dobj.private;
+	struct snd_soc_dapm_path *p;
+
+	if (!widget_ops[widget->id].ipc_unprepare || !swidget->complete)
+		goto sink_unprepare;
+
+	/* unprepare the source widget */
+	widget_ops[widget->id].ipc_unprepare(swidget);
+	swidget->complete = false;
+
+sink_unprepare:
+	/* unprepare all widgets in the sink paths */
+	snd_soc_dapm_widget_for_each_sink_path(widget, p) {
+		if (!p->walking && p->sink->dobj.private) {
+			p->walking = true;
+			sof_unprepare_widgets_in_path(sdev, p->sink);
+			p->walking = false;
+		}
+	}
+}
+
+static int
+sof_prepare_widgets_in_path(struct snd_sof_dev *sdev, struct snd_soc_dapm_widget *widget,
+			    struct snd_sof_platform_stream_params *platform_params,
+			    struct snd_pcm_hw_params *runtime_params,
+			    struct snd_pcm_hw_params *input_params, int dir)
+{
+	const struct sof_ipc_tplg_ops *ipc_tplg_ops = sdev->ipc->ops->tplg;
+	const struct sof_ipc_tplg_widget_ops *widget_ops = ipc_tplg_ops->widget;
+	struct snd_sof_widget *swidget = widget->dobj.private;
+	struct snd_soc_dapm_path *p;
+	int ret;
+
+	if (!widget_ops[widget->id].ipc_prepare || swidget->complete)
+		goto sink_prepare;
+
+	/* prepare the source widget */
+	ret = widget_ops[widget->id].ipc_prepare(swidget, platform_params, runtime_params,
+					     input_params, dir);
```

Comment:
It is unexpected (by naming) that the `ipc_prepare` callback would modify the `input_params`, afaik it is not touching the 'runtime_params' (should this be called 'params'?).

How this works? `ipc_prepare` modifies the second params struct based on the first params _and_ what was passed via the second params? Kind of in place refine/update?
Is the `input_params` is some sort of params for the pipeline thing?

Reply:
@ujfalusi since we are walking the widgets in order from source to sink. The first widget uses the runtime_params and modifes the input_params for its sink based on what it supports (coming from topology). For example: a copier connected to a mixin, will modify the input_params to 32-bit sample size because that the only one the mixin supports

Reply:
Right, but when you move to the next widget in the path, you still pass the `params` and the `input_params` in a same order.

How I see is: the first params is the _input_ params fort he given widget, the second is what will come out from the widget, so the _output_ params of the widget.
The next widget in the path would get the _output_ params of the previous widget as it's _intput_ params and produces an _output_ params. The next widget would again get the previous widget's _output_ params as it's _input_ params and produces the _output_ params.

But the code always gives the original params to all widget in the path and passing along the very same input_params struct for them to modify.

Reply:
@ujfalusi thas one way to look at it but the reason why I named it runtime_params in the first place is this: Take this pipeline for ex: host_copier -> mixin->mixout->dai_copier. The host copier looks at the "params" to decide what its config should be. But it modifies the input_params to set the sample size to 32-bits for the mixin. Then the mixin uses the input_params to decide what its config should be. and passes the input_params unmodified to the mixout. Then mixout picks its config and pass the input_params unmodified to the dai_copier. The dai_copier does not look at the input_params, but choose its config based on the "params". So if the params asked for 16-bits, it would do the conversion and set the input to 32-bits because thats what mixout provides and output to 16-bits because thats what the pcm params is asking for.

Reply:
so in essence, each widget gets 2 params, 1 is the unmodified PCM params and the other one is the input_params as modified by the previous widget in the path.

Reply:
@ranj063, thanks fir the explanation!
Basically you are passing the fw_params and a pipeline_params around and modify the pipeline_params once at the start (host_copier for playback)?
Why do you convert it back at the end?
`host_copier` -> `mixin` -> `mixout` -> `dai_copier`
`16bit` -> `32bit` -> `32bit` -> `16bit` (it gets even worst if we have 24bit)
Why not just:
`16bit` -> `32bit` -> `32bit` -> `32bit`

I know, there must be a good reason...

The naming and function of these `params` are most confusing. For the `host_copier` the first params is the _input_ or host_params or fe_params (the stream which is coming in from user space), the second params is the _output_ params of the `host_copier`, iow, what is going to be passed along the pipeline (if I understand it right).

The runtime_params might be better off to be named host or fe params?

If I get it right, this prepare will be only used by IPC4, but the 'generic' implementation is really a mirror on what IPC4 needs.
At the end you have modules with source and sink, right? you connect the sink of a module to the source of another module.
host_copier
source: host_params/fw_params
sink: host_sink_params

mixin
source: host_sink_params
sink: mixin_sink_params (basically copy of host_sink_params)

mixout
source: mixin_sink_params
sink: mixout_sink_params ( == mixin_sink_params == host_sink_params)

dai_copier
source: mixout_sink_params
sink: host_params/fw_params (? why we do another back conversion?)

Isn't this is what happening in a generic way? Should the code reflect a generic implementation instead of what IPC4 needs?

Reply:
@ranj063 The host application can require a format on the host side only, i.e. a PCM device. What happens in firmware and hardware is not known to the application, e.g. there could be resampling and mixing.

Reply:
@plbossart if the mixing is going to change the format to 32-bit, shoudnt there be a be_dai_link_fixup() for this case to switch the format to 32-bits as well?

Reply:
the problem is that the mixing is not represented except in vague terms. The only place where you can influence format changes is with the DAI in the topology.

Reply:
we could set the format for the DAI copier in topology. But if it is capable of changing the format back to 16-bit, do we really need to do a fixup?

Reply:
why would we go back to 16 bits? It's better to keep the highest possible format at the DAI level, and let the codec deal with it.

Reply:
@plbossart @ujfalusi I've changed the logic in the IPC4 patches to not convert back to 16-bits for playback but we still need the 2 params, pcm params and input_params, because for capture, the host_copier  needs to convert based on pcm params

Reply:
@ranj063, I still find the `input_params` naming a bit confusing. Yes, it is kind of that, but it is not. Especially in case of capture, it is not an input params at all.
The first params is clear: it is the host/fw/pcm params
The second is widget params, perhaps? It might be less confusing if you would not copy the host params to it, but zero it out and let the first widget to fill it up? IN that case it can be seen as the morphing param throughout the pipeline from source's sink to destination's source.
Perhaps?

Reply:
@ujfalusi why is it not input_params for capture? EVery widget has input/putput audio format irrespective of direction. So, input_params is whats being fed to a widget from the previous widget ie the prev widget's output params

Reply:
@ranj063, yes, we don't have disagreement on the function of the second params struct. It is just that (in my understanding) it is to get the output params of the first widget in the pipeline and from there it is going to be passed to the next one as input params and that one can modify it (I know, IPC4 is not doing that) so the next widget will be fed with the output params of the previous.
For me when you kick the `sof_unprepare_widgets_in_path()` chain to motion the second params is not input (it can not be an input on the capture case) it is just a placeholder to drag around to carry the params down the line.

This is probably my nitpicking only, if no one else have similar view, I'm fine with naming it as input_params.

Reply:
@ujfalusi if you have a suggestion for a more intuitive naming, im all ears. btw sof_unprepare_widgets_in_path() does not using params at all

Reply:
source_params?

Reply:
source_params is good for me @ujfalusi are you OK with it too?

Reply:
@plbossart @ujfalusi renamed input_params to source_params now. Let me know what you think about this one